src/Confluent.Kafka/Consumer.cs

```diff
@@ -110,863 +246,400 @@ public class Consumer<TKey, TValue> : IDisposable
-            if (defaultTopicConfig != null)
+            try
             {
-                defaultTopicConfig.ToList().ForEach(
-                    (kvp) => {
-                        if (kvp.Value == null) throw new ArgumentException($"'{kvp.Key}' configuration parameter in 'default.topic.config' must not be null.");
-                        configHandle.Set(kvp.Key, kvp.Value.ToString());
+                var msg = Util.Marshal.PtrToStructureUnsafe<rd_kafka_message>(msgPtr);
+
+                string topic = null;
+                if (this.enableTopicNamesMarshaling)
+                {
+                    if (msg.rkt != IntPtr.Zero)
+                    {
+                        topic = Util.Marshal.PtrToStringUTF8(Librdkafka.topic_name(msg.rkt));
                     }
-                );
-            }
+                }
 
-            // Explicitly keep references to delegates so they are not reclaimed by the GC.
-            rebalanceDelegate = RebalanceCallback;
-            commitDelegate = CommitCallback;
-            errorDelegate = ErrorCallback;
-            logDelegate = LogCallback;
-            statsDelegate = StatsCallback;
+                if (msg.err == ErrorCode.Local_PartitionEOF)
+                {
+                    OnPartitionEOF?.Invoke(this, new TopicPartitionOffset(topic, msg.partition, msg.offset));
+                    return null;
+                }
 
-            IntPtr configPtr = configHandle.DangerousGetHandle();
+                long timestampUnix = 0;
+                IntPtr timestampType = (IntPtr)TimestampType.NotAvailable;
+                if (enableTimestampMarshaling)
+                {
+                    timestampUnix = Librdkafka.message_timestamp(msgPtr, out timestampType);
+                }
+                var timestamp = new Timestamp(timestampUnix, (TimestampType)timestampType);
 
-            LibRdKafka.conf_set_rebalance_cb(configPtr, rebalanceDelegate);
-            LibRdKafka.conf_set_offset_commit_cb(configPtr, commitDelegate);
+                Headers headers = null;
+                if (enableHeaderMarshaling)
+                {
+                    headers = new Headers();
+                    Librdkafka.message_headers(msgPtr, out IntPtr hdrsPtr);
+                    if (hdrsPtr != IntPtr.Zero)
+                    {
+                        for (var i=0; ; ++i)
+                        {
+                            var err = Librdkafka.header_get_all(hdrsPtr, (IntPtr)i, out IntPtr namep, out IntPtr valuep, out IntPtr sizep);
+                            if (err != ErrorCode.NoError)
+                            {
+                                break;
+                            }
+                            var headerName = Util.Marshal.PtrToStringUTF8(namep);
+                            var headerValue = new byte[(int)sizep];
+                            Marshal.Copy(valuep, headerValue, 0, (int)sizep);
```

Comment:
Header values can be NULL, and that should be exposed in the .NET API as well. I don't think this does that

Reply:
well spotted. none of the header functionality supports null values. adding this throughout + integration test.